OpenRA.Mods.Common/Traits/BotModules/Squads/States/GroundStates.cs

```diff
@@ -68,57 +84,115 @@ public void Tick(Squad owner)
 
-			if (ownUnits.Count < owner.Units.Count)
+			var targetActor = ThreatScan(owner, pathGuider, attackScanRadius);
+			if (targetActor != null)
 			{
-				// Since units have different movement speeds, they get separated while approaching the target.
-				// Let them regroup into tighter formation.
-				owner.Bot.QueueOrder(new Order("Stop", leader, false));
-				foreach (var unit in owner.Units.Where(a => !ownUnits.Contains(a)))
-					owner.Bot.QueueOrder(new Order("AttackMove", unit, Target.FromCell(owner.World, leader.Location), false));
+				owner.TargetActor = targetActor;
+				owner.FuzzyStateMachine.ChangeState(owner, new GroundUnitsAttackState(), true);
+				return;
 			}
-			else
+
+			// 2. Force scattered for pathGuider if needed
+			if (StuckInPath <= 0)
 			{
-				var enemies = owner.World.FindActorsInCircle(leader.CenterPosition, WDist.FromCells(owner.SquadManager.Info.AttackScanRadius))
-					.Where(owner.SquadManager.IsPreferredEnemyUnit);
-				var target = enemies.ClosestTo(leader.CenterPosition);
-				if (target != null)
+				if (TryMakeWay > 0)
 				{
-					owner.TargetActor = target;
-					owner.FuzzyStateMachine.ChangeState(owner, new GroundUnitsAttackState(), true);
+					owner.Bot.QueueOrder(new Order("AttackMove", pathGuider, Target.FromCell(owner.World, owner.TargetActor.Location), false));
+
+					foreach (var a in owner.Units)
+					{
+						if (a != pathGuider)
+							owner.Bot.QueueOrder(new Order("Scatter", a, false));
+					}
+
+					TryMakeWay--;
 				}
 				else
-					foreach (var a in owner.Units)
-						owner.Bot.QueueOrder(new Order("AttackMove", a, Target.FromCell(owner.World, owner.TargetActor.Location), false));
+				{
+					// When going through is over, restore the check
+					StuckInPath = StuckInPathCheckTimes + MakeWayTick;
+					TryMakeWay = MakeWayTick;
+				}
+
+				return;
 			}
 
-			if (ShouldFlee(owner))
-				owner.FuzzyStateMachine.ChangeState(owner, new GroundUnitsFleeState(), true);
+			// 3. Check if the squad is stucked due to the map has very twisted path
+			// or currently bridge and tunnel from TS mod
+			if ((pathGuider.CenterPosition - LastPos).LengthSquared <= 4)
+				StuckInPath--;
+			else
+				StuckInPath = StuckInPathCheckTimes;
+
+			LastPos = pathGuider.CenterPosition;
+
+			// 4. Since units have different movement speeds, they get separated while approaching the target.
+
+			/* Let them regroup into tighter formation towards "pathGuider".
+			 *
+			 * "unitsArea" means the space the squad units will occupy (if 1 per Cell).
+			 * pathGuider only stop when scope of "unitsAround" is not covered all units;
+			 * units in "unitsHurryUp"  will catch up,
+			 * which keep the team tight while not stucked.
+			 *
+			 * Imagining "unitsArea" takes up a a place shape like square, we need to draw a circle
+			 * to cover the the enitire circle.
+			 *
+			 * When look around, pathGuider find units around to decide if it need to continue.
```

Comment:
How about: "pathGuide will check how many squad members are around to decide if it needs to continue" ?

Reply:
Good one, fixed